### PR TITLE
[WIP] Add draft for caching

### PR DESCRIPTION
**Pull request by jainxrohit**
These changes are not complete.
Pending items:
   1. Naming
    2. package refactoring
    3. Tests

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -62,7 +62,7 @@
         <dep.nexus-staging-plugin.version>1.6.8</dep.nexus-staging-plugin.version>
         <dep.asm.version>6.2.1</dep.asm.version>
         <dep.gcs.version>1.9.17</dep.gcs.version>
-        <dep.alluxio.version>2.1.2</dep.alluxio.version>
+        <dep.alluxio.version>2.3.0-SNAPSHOT</dep.alluxio.version>
```
Comment (highker) [inline pom.xml]: Do they have a stable release now? It will be easy to review if they have a release.

Comment (jainxrohit) [reply pom.xml]: In a day or two, alluxio will cut the release.

File: presto-cache/src/main/java/com/facebook/presto/cache/alluxio/CacheFactory.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package com.facebook.presto.cache.alluxio;
+
+import alluxio.conf.AlluxioConfiguration;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+
+import java.net.URI;
+
+public interface CacheFactory
```
Comment (highker) [inline presto-cache/src/main/java/com/facebook/presto/cache/alluxio/CacheFactory.java]: We might not need this interface or abstraction if this is only for Alluxio. Just call it AlluxioCacheFactory

File: presto-hive/src/main/java/com/facebook/presto/hive/cache/HiveCachingHdfsConfiguration.java [hunk 3]
```diff
@@ -59,12 +65,17 @@ public Configuration getConfiguration(HdfsContext context, URI uri)
         @SuppressWarnings("resource")
         Configuration config = new CachingJobConf((factoryConfig, factoryUri) -> {
             try {
+                FileSystem fileSystem = (new Path(factoryUri)).getFileSystem(hiveHdfsConfiguration.getConfiguration(context, factoryUri));
+                FileSystem dataTierFileSystem = fileSystem;
+                if (cacheConfig.isCachingEnabled() && cacheConfig.getCacheType() == ALLUXIO) {
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/cache/HiveCachingHdfsConfiguration.java]: We don't need to wrap AlluxioFS into CachingFS. AlluxioFS and CachingFS are both caching FS solutions. So just return AlluxioFS should be good enough. If `getCacheType() == LOCAl_RANGE`, then we return cachingFS.

File: presto-hive/src/main/java/com/facebook/presto/hive/cache/alluxio/AlluxioCachingModule.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+package com.facebook.presto.hive.cache.alluxio;
+
+import com.facebook.airlift.configuration.AbstractConfigurationAwareModule;
+import com.facebook.presto.hive.DynamicConfigurationProvider;
+import com.google.inject.Binder;
+import com.google.inject.Scopes;
+
+import static com.google.inject.multibindings.Multibinder.newSetBinder;
+
+public class AlluxioCachingModule
```
Comment (highker) [inline presto-hive/src/main/java/com/facebook/presto/hive/cache/alluxio/AlluxioCachingModule.java]: We can make it a CacheModule for presto-hive. Put all cache-related binding to this module including alluxio and local-range

File: presto-cache/src/main/java/com/facebook/presto/cache/alluxio/CacheFactory.java (hunk 2 above)
Comment (jainxrohit) [reply presto-cache/src/main/java/com/facebook/presto/cache/alluxio/CacheFactory.java]: This factory is required for both Hive and Prism. And it produces different filesystems for Hive and Prism. So I think we would need the interface. If we want to call the Interface AlluxioCacheFactory, then we can discuss.
I was thinking if we build another filesystem or memory based caching and decided to use this pattern then this factory would be useful.

File: presto-hive/src/main/java/com/facebook/presto/hive/cache/HiveCachingHdfsConfiguration.java (hunk 3 above)
Comment (jainxrohit) [reply presto-hive/src/main/java/com/facebook/presto/hive/cache/HiveCachingHdfsConfiguration.java]: Caching FS is the abstraction for both Alluxio and LOCAL_RANGE. The real caching solution will be determined by the datatierFileSystem. We can talk about this if it is not clear.

File: presto-hive/src/main/java/com/facebook/presto/hive/cache/alluxio/AlluxioCachingModule.java (hunk 4 above)
Comment (jainxrohit) [reply presto-hive/src/main/java/com/facebook/presto/hive/cache/alluxio/AlluxioCachingModule.java]: Sounds good, I will try to see how this fits in.

**Pull request closed without merging**